### PR TITLE
Alternative `ModelIO` instance for the reference implementation

### PR DESCRIPTION
**Pull request by jorisdral**

### CODE REVIEW

**Comment by dcoutts** [comment]
A bit of explanation of motivation would be good here.

**Comment by mheinzel** [comment]
A refactoring to reduce code duplication. Previously, multiple models (table-level, session-level, mutable API on top) had been implemented twice, once for normal tables and once for monoidal tables.

The main change is that now the table- and session-level don't make a distinction between normal and monoidal tables any more. This also comes with some other re-organisation, which IMO makes the structure and dependencies between different models and tests clearer.

The are now three models, each depending on the previous one:
- `Model.Table`: a pure model of a single table (e.g. `updates :: [(k, Update v b)] -> Table k v b -> Table k v b`)
- `Model.Session`: a pure model of a session (e.g. `updates :: MonadState Model m => [(k, Update v b)] -> TableHandle k v b -> m ()`)
- `Model.IO.{Normal,Monoidal}`: STM-based models of an explicit (potentially closed) session (e.g. `updates :: MonadSTM m => MSession m -> [(k, Update v b)] -> TableHandle k v b -> m ()`), separated between normal/monoidal, which requires some type conversions such as `convLookupResult`

In a way, the `StateMachine` tests are a fourth model, based on `Model.Session`.


@jorisdral, since I picked up the work from you, I left the original commit in place (only rebased it), so you can see what changed from when it was handed over. A few things I changed:

- moved modules around to have all models under a `Model` directory, each with a small description.
- re-added export lists
- avoided `convLookupResult` creeping into `observeReal` etc., and the need for `convLookupResult'` etc. in general.
- avoided `ResolveValue` creeping into any `Normal` modules by passing `ResolveSerialisedValue` as an argument instead of using a class in `Model.Session`

File: test/Database/LSMTree/Class/Normal.hs [hunk 1]
```diff
@@ -169,55 +213,126 @@ withTableOpen ::
...
+     )
   => Maybe k
   -> h m k v blob
   -> (Cursor h m k v blob -> m a)
   -> m a
 withCursor offset hdl = bracket (newCursor offset hdl) (closeCursor (Proxy @h))
 
 {-------------------------------------------------------------------------------
-  Model instance
+  Model 2 instance
```
[the remark this answers is not in the document]
Comment (mheinzel) [reply test/Database/LSMTree/Class/Normal.hs]: I added a short module description for each model, but it could probably still need some more documentation.

> What I want to see somewhere in these models and instances is a high level explanation of them all.

Not sure where best to put that, but it could be make use of the comparison in https://github.com/IntersectMBO/lsm-tree/pull/403#issuecomment-2426564132.

File: test/Database/LSMTree/Class/Monoidal.hs [hunk 2]
```diff
@@ -12,39 +14,27 @@ module Database.LSMTree.Class.Monoidal (
...
 import qualified Database.LSMTree.Monoidal as R
-import qualified Database.LSMTree.SessionModel as M2
-import qualified Database.LSMTree.TableModel as M22
 
+-- | Model-specific constraints
+type C k v blob = (C_ k, C_ v, C_ blob)
+type C_ a = (Show a, Eq a, Typeable a)
 
--- | Class abstracting over table handle operations.
+-- | Class abstracting over session operations.
```
Comment (jorisdral) [inline test/Database/LSMTree/Class/Monoidal.hs:37]: It's table operations, not session operations

File: test/Database/LSMTree/Class/Monoidal.hs [hunk 3]
```diff
@@ -12,39 +14,27 @@ module Database.LSMTree.Class.Monoidal (
...
-  , convQueryResult
-  , convQueryResult'
-  , convUpdate
-  , convUpdate'
-  , MSession (..)
-  , MTableHandle (..)
-  , MCursor (..)
-  , MErr (..)
-  , MBlobRef (..)
+  , module Types
```
Comment (jorisdral) [inline test/Database/LSMTree/Class/Monoidal.hs:17]: I'm not a big fan of re-exports, because it makes it hard to follow where identifiers are coming from. However, it's also weird that one would have to import a bunch of other modules. What do you think?

If we change this, we should make sure it's consistent across the model/class modules

I'm currently leaning towards having no re-exports in the `Class` and `Model` modules. The model/class hierarchy is already a little bit complicated, and re-rexports of identifiers might not help

Also, I haven't really been consistent with re-exports vs. no re-exports in these modules myself, so I'm mainly criticising myself :stuck_out_tongue_closed_eyes:

File: test/Test/Database/LSMTree/Class/Normal.hs [hunk 4]
```diff
@@ -20,14 +20,11 @@ import           Data.Word (Word64)
...
                      withTableNew, withTableOpen)
 import qualified Database.LSMTree.Class.Normal as Class
-import           Database.LSMTree.Common (IOLike, Labellable (..), Range (..),
-                     SerialiseKey, SerialiseValue, mkSnapshotName)
+import           Database.LSMTree.Common (mkSnapshotName)
 import           Database.LSMTree.Extras.Generators ()
-import           Database.LSMTree.Monoidal (ResolveValue (..))
-import           Database.LSMTree.Normal (LookupResult (..), QueryResult (..),
-                     Update (..))
+import qualified Database.LSMTree.Model.IO.Normal as Model
```
Comment (jorisdral) [inline test/Test/Database/LSMTree/Class/Normal.hs:25]: `as ModelIO`

File: test/Test/Database/LSMTree/Class/Monoidal.hs [hunk 5]
```diff
@@ -15,14 +15,12 @@ import           Data.Word (Word64)
...
                      withTableNew, withTableOpen)
 import qualified Database.LSMTree.Class.Monoidal as Class
-import           Database.LSMTree.Common (Labellable (..), mkSnapshotName)
+import           Database.LSMTree.Common (mkSnapshotName)
 import           Database.LSMTree.Extras.Generators ()
-import           Database.LSMTree.Monoidal (IOLike, LookupResult (..),
-                     QueryResult (..), Range (..), ResolveValue (..),
-                     SerialiseKey, SerialiseValue, Update (..),
-                     resolveDeserialised)
+import qualified Database.LSMTree.Model.IO.Monoidal as Model
```
Comment (jorisdral) [inline test/Test/Database/LSMTree/Class/Monoidal.hs:20]: `as ModelIO`

File: test/Database/LSMTree/Model/IO/Monoidal.hs [hunk 6]
```diff
@@ -0,0 +1,90 @@
+{-# LANGUAGE TypeFamilies #-}
+
+-- | A @TVar@-based monoidal model of a potentially closed mutable session,
+-- giving an instance of `Class.IsTableHandle`.
```
Comment (jorisdral) [inline test/Database/LSMTree/Model/IO/Monoidal.hs:4]: I don't think it's too important that it is using a `TVar`. I would emphasise that this is the pure `Session` model lifted into `IO`

File: test/Database/LSMTree/Model/IO/Normal.hs [hunk 7]
```diff
@@ -0,0 +1,109 @@
+{-# LANGUAGE TypeFamilies #-}
+
+-- | A @TVar@-based normal (i.e. non-monoidal) model of a potentially closed
+-- mutable session, giving an instance of `Class.IsTableHandle`.
```
Comment (jorisdral) [inline test/Database/LSMTree/Model/IO/Normal.hs:4]: I don't think it's too important that it is using a `TVar`. I would emphasise that this is the pure `Session` model lifted into `IO`

File: test/Database/LSMTree/Model/IO/Normal.hs [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+import           Control.Exception (Exception)
+import           Control.Monad.Class.MonadThrow (MonadThrow (..))
+import qualified Database.LSMTree.Class.Normal as Class
+import qualified Database.LSMTree.Model.Session as Model
+import           NoThunks.Class (AllowThunk (..), NoThunks)
+
+newtype Session m = Session (StrictTVar m (Maybe Model.Model))
+
+deriving via AllowThunk (Session IO)
+    instance NoThunks (Session IO)
```
Comment (jorisdral) [inline test/Database/LSMTree/Model/IO/Normal.hs:29]: I'd keep this in the `StateMachine` tests, even if it is an orphan

File: test/Database/LSMTree/Model/Session.hs [hunk 9]
```diff
@@ -1,8 +1,73 @@
-{-# LANGUAGE TypeFamilies #-}
-
-{-# OPTIONS_GHC -Wno-missing-export-lists #-}
-
-module Database.LSMTree.SessionModel where
+-- | A pure model of a session, supporting both blobs and mupserts.
+--
+-- Builds on top of "Database.LSMTree.Model.Table", adding table handle and
+-- snapshot administration.
```
Comment (jorisdral) [inline test/Database/LSMTree/Model/Session.hs:4]:
```suggestion
-- | A pure model of a single session containing multiple tables.
--
-- This model supports all the features for /both/ normal and monoidal tables. For example, the session model supports both blobs and mupserts. The former is typically only provided by the normal API, and the latter only by the monoidal API. 
--
-- The session model builds on top of "Database.LSMTree.Model.Table", adding table handle and
-- snapshot administration.
```

File: test/Database/LSMTree/Model/Table.hs [hunk 10]
```diff
@@ -1,8 +1,46 @@
 {-# LANGUAGE TypeFamilies #-}
 
-{-# OPTIONS_GHC -Wno-missing-export-lists #-}
-
-module Database.LSMTree.TableModel where
+-- | A pure model of a single table, supporting both blobs and mupserts.
```
Comment (jorisdral) [inline test/Database/LSMTree/Model/Table.hs:3]:
```suggestion
-- | A pure model of a single table.
--
-- This model supports all the features for /both/ normal and monoidal tables. For example, the table model supports both blobs and mupserts. The former is typically only provided by the normal API, and the latter only by the monoidal API. 
```

File: test/Database/LSMTree/Model/IO/Monoidal.hs
Comment (jorisdral) [inline test/Database/LSMTree/Model/IO/Monoidal.hs:1]: If we're bringing back an IO model module, let's put it in `Database.LSMTree.ModelIO.Monoidal` instead.

File: test/Database/LSMTree/Model/IO/Normal.hs
Comment (jorisdral) [inline test/Database/LSMTree/Model/IO/Normal.hs:1]: If we're bringing back an IO model module, let's put it in `Database.LSMTree.ModelIO.Normal` instead.

File: test/Test/Database/LSMTree/Model/Table.hs [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+updates = Model.updates Model.getResolve
+
+inserts :: V.Vector (Key, Value, Maybe Blob) -> Table Key Value Blob -> Table Key Value Blob
+inserts = Model.inserts Model.getResolve
+
+deletes :: V.Vector Key -> Table Key Value Blob -> Table Key Value Blob
+deletes = Model.deletes Model.getResolve
+
+mupserts :: V.Vector (Key, Value) -> Table Key Value Blob -> Table Key Value Blob
+mupserts = Model.mupserts Model.getResolve
```
Comment (jorisdral) [inline test/Test/Database/LSMTree/Model/Table.hs:55]: Without the `ResolveValue -> ResolveSerialisedValue` change, we would not need these functions

**Review by jorisdral — COMMENTED** [review]
Thanks for updating the PR! Few things to address, and then we can merge :rocket: 

> ...
> 
> The are now three models, each depending on the previous one:
> 
>     * `Model.Table`: a pure model of a single table (e.g. `updates :: [(k, Update v b)] -> Table k v b -> Table k v b`)
> 
>     * `Model.Session`: a pure model of a session (e.g. `updates :: MonadState Model m => [(k, Update v b)] -> TableHandle k v b -> m ()`)
> 
>     * `Model.IO.{Normal,Monoidal}`: STM-based models of an explicit (potentially closed) session (e.g. `updates :: MonadSTM m => MSession m -> [(k, Update v b)] -> TableHandle k v b -> m ()`), separated between normal/monoidal, which requires some type conversions such as `convLookupResult`
> 
> 
> In a way, the `StateMachine` tests are a fourth model, based on `Model.Session`.

You could put this overview of the hierarchy, maybe extended to also metion the class modules, in a top-level module that only contains a bit of documenation. For example, a new `Database.LSMTree.ModelHierarchy` module, or something similar. 

Also, isn't the `StateMachine` model "just" the session model? I wouldn't say it adds any functionality over the session model, it is just adding some boilerplate for the lockstep framework

> @jorisdral, since I picked up the work from you, I left the original commit in place (only rebased it), so you can see what changed from when it was handed over. A few things I changed:
> 
>     * moved modules around to have all models under a `Model` directory, each with a small description.
> 
>     * re-added export lists
> 
>     * avoided `convLookupResult` creeping into `observeReal` etc., and the need for `convLookupResult'` etc. in general.
> 
>     * avoided `ResolveValue` creeping into any `Normal` modules by passing `ResolveSerialisedValue` as an argument instead of using a class in `Model.Session`

Regarding avoiding `convLookupResult`, what it the concrete change that made this possible?

Regarding `ResolveValue`, I don't think it was "bad" to have it in the normal modules

File: test/Database/LSMTree/Model/Session.hs [hunk 12]
```diff
@@ -1,26 +1,6 @@
...
-
--- | Model a single session, allowing multiple tables.
---
--- Builds on top of "Database.LSMTree.Model.Normal", adding table handle and
--- snapshot administration.
-module Database.LSMTree.Model.Normal.Session (
-    -- * Model
+-- | A pure model of a session, supporting both blobs and mupserts.
+module Database.LSMTree.Model.Session (
+  -- * Model
```
[the remark this answers is not in the document]
Comment (mheinzel) [reply test/Database/LSMTree/Model/Session.hs]: @jorisdral Do you think we should rename it?

File: test/Database/LSMTree/Model/IO/Monoidal.hs
Comment (mheinzel) [reply test/Database/LSMTree/Model/IO/Monoidal.hs:1]: I thought it made sense to put the different kinds of models under a shared `Model` hierarchy. Then the documentation about the different kinds of models could maybe even just go into the `Model` module (which is otherwise empty).

But no strong feelings on this one.

File: test/Test/Database/LSMTree/Model/Table.hs (hunk 11 above)
Comment (mheinzel) [reply test/Test/Database/LSMTree/Model/Table.hs:55]: Yeah, it adds a little bit of boilerplate here (didn't want to make the properties themselves more noisy), but I think it's overall still nicer than having spurious `ResolveValue` constraints show up.

File: test/Database/LSMTree/Model/IO/Monoidal.hs
Comment (jorisdral) [reply test/Database/LSMTree/Model/IO/Monoidal.hs:1]: Let's do your suggestion

**Comment by mheinzel** [comment]
Thanks for the review, good points!

> Regarding avoiding `convLookupResult`, what it the concrete change that made this possible?

```diff
-    MLookupResult :: (Model.C_ v, Model.C_ blob)
-                  => Model.LookupResult v (Val h (WrapBlobRef h IO blob))
-                  -> Val h (R.LookupResult v (WrapBlobRef h IO blob))
+    MLookupResult :: (Class.C_ v, Class.C_ blob)
+                  => LookupResult v (Val h (WrapBlobRef h IO blob))
+                  -> Val h (LookupResult v (WrapBlobRef h IO blob))
```

And same for `OLookupResult`. So the conversion from `Model.LookupResult` (which can contain mupserts *and* blobs) to the real/`Normal` version happens as part of `runModel`.

File: test/Database/LSMTree/Class/Monoidal.hs (hunk 3 above)
Comment (mheinzel) [reply test/Database/LSMTree/Class/Monoidal.hs:17]: I don't think it makes a big difference, but let me share some of my reasoning.

IMHO it's sometimes a bit clearer to have the re-export, for example with `convLookupResult :: Model.LookupResult v b -> Class.LookupResult v b` it makes it clear that it exists because the model's type doesn't quite fit the class we want to implement. If it was something like `-> Real.LookupResult`, the purpose is already less obvious.

Another way of thinking about it: The class uses a lot of the same types as the public API of the implementation. Now the question is: Is that a consciously chosen property of the class? If so, we can expect downstream code to also import something like `LSMTree.Normal`. If it is more of an implementation detail, we should re-export the types and use them as `Class._`, which allows using/implementing the class without considering the implementation, and also makes it easier to swap types out later.

**Comment by mheinzel** [comment]
I rebased and addressed all feedback..

File: test/Test/Database/LSMTree/Class/Normal.hs [hunk 13]
```diff
@@ -35,17 +32,18 @@ import qualified Test.Util.FS as FS
 
 tests :: TestTree
 tests = testGroup "Test.Database.LSMTree.Class.Normal"
-    [ testGroup "Model" $ zipWith ($) (props tbl1) expectFailures1
+    [ testGroup "Model2" $ zipWith ($) (props tbl3) expectFailures3
```
Comment (jorisdral) [inline test/Test/Database/LSMTree/Class/Normal.hs:35]:
```suggestion
    [ testGroup "Model" $ zipWith ($) (props tbl3) expectFailures1
```

File: test/Test/Database/LSMTree/Class/Normal.hs [hunk 14]
```diff
@@ -35,17 +32,18 @@ import qualified Test.Util.FS as FS
...
     , testGroup "Real"  $ zipWith ($) (props tbl2) expectFailures2
     ]
   where
-    tbl1 :: Proxy M.TableHandle
-    tbl1 = Setup {
-          testTableConfig = M.TableConfig
-        , testWithSessionArgs = \action -> action NoSessionArgs
+
+    tbl3 :: Proxy ModelIO.TableHandle
+    tbl3 = Setup {
```
Comment (jorisdral) [inline test/Test/Database/LSMTree/Class/Normal.hs:41]:
```suggestion
    tbl1 :: Proxy ModelIO.TableHandle
    tbl1 = Setup {
```

File: test/Test/Database/LSMTree/Class/Normal.hs [hunk 15]
```diff
@@ -35,17 +32,18 @@ import qualified Test.Util.FS as FS
...
+          testTableConfig = ModelIO.TableConfig
+        , testWithSessionArgs = \action -> action ModelIO.NoSessionArgs
         }
 
-    expectFailures1 = repeat False
+    expectFailures3 = repeat False
```
Comment (jorisdral) [inline test/Test/Database/LSMTree/Class/Normal.hs:46]:
```suggestion
    expectFailures1 = repeat False
```

**Review by jorisdral — COMMENTED** [review]
LGTM! I can't approve, so you'll have to do it @mheinzel

**Review by mheinzel — APPROVED** [review]
